cli/internal/helm/upgrade.go

```diff
@@ -114,53 +115,55 @@ func (c *UpgradeClient) shouldUpgrade(releaseName string, newVersion semver.Semv
 // Upgrade runs a helm-upgrade on all deployments that are managed via Helm.
 // If the CLI receives an interrupt signal it will cancel the context.
 // Canceling the context will prompt helm to abort and roll back the ongoing upgrade.
-func (c *UpgradeClient) Upgrade(ctx context.Context, config *config.Config, idFile clusterid.File, timeout time.Duration, allowDestructive, force bool, upgradeID string) error {
+func (c *UpgradeClient) Upgrade(ctx context.Context, config *config.Config, idFile clusterid.File, timeout time.Duration, allowDestructive, force bool, upgradeID string, conformance bool, helmWaitMode WaitMode, masterSecret uri.MasterSecret, serviceAccURI string, validK8sVersion versions.ValidK8sVersion, output terraform.ApplyOutput) error {
 	upgradeErrs := []error{}
-	upgradeReleases := []*chart.Chart{}
-	newReleases := []*chart.Chart{}
-	for _, info := range getManagedCharts(config) {
-		c.log.Debugf("Checking release %s", info.releaseName)
-		chart, err := loadChartsDir(helmFS, info.path)
-		if err != nil {
-			return fmt.Errorf("loading chart: %w", err)
-		}
+	upgradeReleases := []Release{}
+	newReleases := []Release{}
+
+	clusterName := clusterid.GetClusterName(config, idFile)
+	helmLoader := NewLoader(config.GetProvider(), validK8sVersion, clusterName)
+	c.log.Debugf("Created new Helm loader")
 
+	releases, err := helmLoader.LoadReleases(config, conformance, helmWaitMode, masterSecret, masterSecret.Salt, serviceAccURI, idFile, output)
```

Comment:
This function uses the wrong salt for `measurementSalt`.

We currently don't save this salt (generated during init) anywhere, since it was not needed at the time as we didn't update the join-config ConfigMap using Helm.
I suggest saving the measurementSalt during init into the cluster IDs file, and then load it during upgrade apply.

Reply:
fixed it, thanks. type safety would be helpful to inhibit misusage